src/record_ts.h

```diff
@@ -0,0 +1,873 @@
+            //
+            CallData *call = nullptr;
+            for (CallData *tmp : calls_assembled) {
+                if (tmp->offset == dst) {
+                    call = tmp;
+                    break;
+                }
+            }
+            if (call) {
+                capture_call_offset(dst, size);
+                jitc_log(LogLevel::Debug, "    captured call offset");
+            }
+        }
+    }
+
+    /// Sum over elements within blocks
+    void block_reduce(VarType vt, ReduceOp op, uint32_t size,
+                      uint32_t block_size, const void *in, void *out) override {
+        if (!paused()) {
+            try {
+                record_block_reduce(vt, op, size, block_size, in, out);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->block_reduce(vt, op, size, block_size, in,
+                                              out);
+    }
+
+    void block_prefix_reduce(VarType vt, ReduceOp op, uint32_t size,
+                             uint32_t block_size, bool exclusive, bool reverse,
+                             const void *in, void *out) override {
+        if (!paused()) {
+            try {
+                record_block_prefix_reduce(vt, op, size, block_size, exclusive,
+                                           reverse, in, out);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->block_prefix_reduce(
+            vt, op, size, block_size, exclusive, reverse, in, out);
+    }
+
+    /// Compute a dot product of two equal-sized arrays
+    void reduce_dot(VarType type, const void *ptr_1, const void *ptr_2,
+                    uint32_t size, void *out) override {
+        jitc_log(
+            LogLevel::Warn,
+            "RecordThreadState::reduce_dot(): unsupported function recording!");
+        scoped_pause();
+        return this->m_internal->reduce_dot(type, ptr_1, ptr_2, size, out);
+    }
+
+    /// Asynchronously update a single element in memory
+    void poke(void *dst, const void *src, uint32_t size) override {
+        jitc_log(LogLevel::Warn,
+                 "RecordThreadState::poke(): unsupported function recording!");
+        scoped_pause();
+        return this->m_internal->poke(dst, src, size);
+    }
+
+    void aggregate(void *dst, AggregationEntry *agg, uint32_t size) override {
+        if (!paused()) {
+            try {
+                record_aggregate(dst, agg, size);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        this->m_internal->aggregate(dst, agg, size);
+    }
+
+    // Enqueue a function to be run on the host once backend computation is done
+    void enqueue_host_func(void (*callback)(void *), void *payload) override {
+        scoped_pause();
```

Comment:
Should this have a warning / exception that `enqueue_host_func` isn't recorded? Or is it safe to have it not recorded and we can still replay later without issues?

Reply:
As far as I can tell, this function is only called when freeing numpy arrays, which should be fine during recording.